File: Ahorn/entities/templegateReversed.jl
module TempleGateReversed

using ..Ahorn, Maple

@mapdef Entity "DJMapHelper/templeGateReversed" TempleGate(x::Integer, y::Integer, height::Integer=48, type::String="CloseBehindPlayer", sprite::String="default")

const placements = Ahorn.PlacementDict(
    "Temple Gate Reversed (DJMapHelper)" => Ahorn.EntityPlacement(
        TempleGate
    )
)

textures = String["default", "mirror", "theo"]

const temple_gate_modes = String[
    "CloseBehindPlayer",
    "CloseBehindPlayerAlways",
    "HoldingTheo",
    "CloseBehindPlayerAndTheo"
]

Ahorn.editingOptions(entity::TempleGate) = Dict{String, Any}(
    "type" => temple_gate_modes,
    "sprite" => textures
)

function Ahorn.selection(entity::TempleGate)
    x, y = Ahorn.position(entity)
    height = Int(get(entity.data, "height", 8))

    return Ahorn.Rectangle(x - 4, y, 15, height)
end

sprites = Dict{String, String}(
    "default" => "objects/door/TempleDoor00",
    "mirror" => "objects/door/TempleDoorB00",
    "theo" => "objects/door/TempleDoorC00"
)

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::TempleGate, room::Maple.Room)
    sprite = get(entity.data, "sprite", "default")

    if haskey(sprites, sprite)
        Ahorn.drawImage(ctx, sprites[sprite], -4, 0)
    end
end

end

File: Loenn/entities/temple_gate_reversed.lua
local drawableSprite = require("structs.drawable_sprite")

local templeGate = {}

templeGate.name = "DJMapHelper/templeGateReversed"
templeGate.depth = -9000
templeGate.canResize = { false, false }
templeGate.placements = {
    name = "normal",
    data = {
        height = 48,
        sprite = "default",
        ["type"] = "CloseBehindPlayer",
    }
}

local textures = {
    default = "objects/door/TempleDoor00",
    mirror = "objects/door/TempleDoorB00",
    theo = "objects/door/TempleDoorC00"
}

local types = {
    "CloseBehindPlayer",
    "CloseBehindPlayerAlways",
    "HoldingTheo",
    "CloseBehindPlayerAndTheo"
}

templeGate.fieldInformation = {
    sprite = {
        fieldType = "anything",
        options = textures,
        editable = false
    },
    ["type"] = {
        fieldType = "anything",
        options = types,
        editable = false
    },
}

function templeGate.sprite(room, entity)
    local variant = entity.sprite or "default"
    local texture = textures[variant] or textures["default"]
    local sprite = drawableSprite.fromTexture(texture, entity)

    -- Weird offset from the code, justifications are from sprites.xml
    sprite:setJustification(0.5, 0.0)
    sprite:addPosition(4, -8)

    return sprite
end

return templeGate